applications/ChatGPT/chatgpt/replay_buffer/detached.py

```diff
@@ -0,0 +1,87 @@
+import torch
+import random
+from typing import List, Any
+from .base import ReplayBuffer
+# from torch.multiprocessing import Queue
+from ray.util.queue import Queue
+import ray
+from chatgpt.experience_maker.base import Experience
+from .utils import BufferItem, make_experience_batch, split_experience_batch
+from threading import Lock
+import copy
+
+class DetachedReplayBuffer:
+    '''
+        Detached replay buffer. Share Experience across workers on the same node. 
+        Therefore a trainer node is expected to have only one instance. 
+        It is ExperienceMakerHolder's duty to call append(exp) method, remotely.
+    
+    Args:
+        sample_batch_size: Batch size when sampling. Exp won't enqueue until they formed a batch.
+        tp_world_size: Number of workers in the same tp group
+        limit: Limit of number of experience sample BATCHs. A number <= 0 means unlimited. Defaults to 0.
+        cpu_offload: Whether to offload experience to cpu when sampling. Defaults to True.
+    '''
+
+    def __init__(self, sample_batch_size: int, tp_world_size: int = 1, limit : int = 0, cpu_offload: bool = True) -> None:
+        self.cpu_offload = cpu_offload
+        self.sample_batch_size = sample_batch_size
+        self.limit = limit
+        self.items = Queue(self.limit)
```

Comment:
what is the use of self.items ?

Reply:
It is the inner buffer, holding batched experience. 
I use ray.util.queue (torch.multiprocessing.queue is also ok) to deal with concurrent IO.